tests/capture/test_capture_diff.py

```diff
@@ -0,0 +1,252 @@
+
+    method = kwargs.get("method", None)
+    h = kwargs.get("h", None)
+    jaxpr = jax.make_jaxpr(func)(0.6)
+    with pytest.raises(ValueError, match=f"'{method=}' and '{h=}' without QJIT"):
+        func(0.6)
+    with pytest.raises(ValueError, match=f"'{method=}' and '{h=}' without QJIT"):
+        jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, 0.6)
+
+
+def test_error_with_non_scalar_function():
+    """Test that an error is raised if the differentiated function has non-scalar outputs."""
+    with pytest.raises(TypeError, match="Grad only applies to scalar-output functions."):
+        jax.make_jaxpr(qml.grad(jnp.sin))(jnp.array([0.5, 0.2]))
+
+
+def grad_eqn_assertions(eqn, argnum=None, n_consts=0):
+    argnum = [0] if argnum is None else argnum
+    assert eqn.primitive == grad_prim
+    assert set(eqn.params.keys()) == {"argnum", "n_consts", "jaxpr", "method", "h"}
+    assert eqn.params["argnum"] == argnum
+    assert eqn.params["n_consts"] == n_consts
+    assert eqn.params["method"] is None
+    assert eqn.params["h"] is None
+
+
+@pytest.mark.parametrize("x64_mode", (True, False))
+@pytest.mark.parametrize("argnum", ([0, 1], [0], [1], 0, 1))
+def test_classical_grad(x64_mode, argnum):
+    """Test that the qml.grad primitive can be captured with classical nodes."""
+
+    initial_mode = jax.config.jax_enable_x64
+    jax.config.update("jax_enable_x64", x64_mode)
+    fdtype = jnp.float64 if x64_mode else jnp.float32
+
+    def inner_func(x, y):
+        return jnp.prod(jnp.sin(x) * jnp.cos(y) ** 2)
+
+    def func_qml(x):
+        return qml.grad(inner_func, argnum=argnum)(x, 0.4 * jnp.sqrt(x))
+
+    def func_jax(x):
+        return jax.grad(inner_func, argnums=argnum)(x, 0.4 * jnp.sqrt(x))
+
+    x = 0.7
+    jax_out = func_jax(x)
+    assert qml.math.allclose(func_qml(x), jax_out)
+
+    # Check overall jaxpr properties
+    if isinstance(argnum, int):
+        argnum = [argnum]
+    jaxpr = jax.make_jaxpr(func_qml)(x)
+    assert jaxpr.in_avals == [jax.core.ShapedArray((), fdtype, weak_type=True)]
+    assert len(jaxpr.eqns) == 3
+    assert jaxpr.out_avals == [jax.core.ShapedArray((), fdtype, weak_type=True)] * len(argnum)
+
+    grad_eqn = jaxpr.eqns[2]
+    grad_eqn_assertions(grad_eqn, argnum=argnum)
+    assert [var.aval for var in grad_eqn.outvars] == jaxpr.out_avals
+    assert len(grad_eqn.params["jaxpr"].eqns) == 6  # 5 numeric eqns, 1 conversion eqn
+
+    manual_eval = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x)
+    assert qml.math.allclose(manual_eval, jax_out)
+
+    jax.config.update("jax_enable_x64", initial_mode)
+
+
+@pytest.mark.parametrize("x64_mode", (True, False))
+def test_nested_grad(x64_mode):
+    """Test that nested qml.grad primitives can be captured.
+    We use the function
+    f(x) = sin(x)^3
+    f'(x) = 3 sin(x)^2 cos(x)
+    f''(x) = 6 sin(x) cos(x)^2 - 3 sin(x)^3
+    f'''(x) = 6 cos(x)^3 - 12 sin(x)^2 cos(x) - 9 sin(x)^2 cos(x)
+    """
+    initial_mode = jax.config.jax_enable_x64
```

Comment:
We could probably put the precision toggling stuff inside a fixture. That way the precision can be automatically parametrized and we don't need to reset the precision manually at the end of each test.

Reply:
The current implementation is also used throughout the rest of the capture tests, so it's by no means needed. It could be worth just doing that in a separate PR for all tests in the `capture` folder.

Reply:
Good idea! I agree that we should do this clean up in a separate PR, to get these ones in swiftly.